### DISCUSSION TITLE
How can we improve the experience for Rust users?

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#6557] [Feature Requests]
I read a few comments on Twitter about how the Rust experience in Zed feels inferior compared to the Rust experience in VS Code.  What sort of Rust-related features / conveniences are you missing Zed that is in VS Code, or another editor using [rust-analyzer](https://rust-analyzer.github.io)?

**Comment by failable** [comment]
Some are not really specific to Rust, but I'm facing when coding Rust.

1. Missing snippets (I hate to type `println!("{:#?}", some_var);` every time, especially that `}` can not be completed automatically in the quotes)
2. Optional to show Inlay hints
3. Can't selecting a struct / impl / fn / block / parens directly 
4. Can't not complete words (RA is too intelligent and only work in strict context)
5. Can't not copy diagnostic messages (have to manual type the message in Google)
6. The diagnostic message can't not be displayed completely without moving the mouse to the point or opening the diagnostic buffer (inconvenient I almost don't use the mouse)
7. The hover doc can not be copied (very often want to try the doc example as a beginner)
8. Can't not comment on black line (annoying when adding docs)
9. Sometimes everything (e.g. completion / diagnostic) stop working until I manual run a `cargo build` and found that I have a typo in `Cargo.toml`. Before that I restarted the LSP many times, no logs were displayed. Things failed silently, like when it happens with `settings.json`.
10. Can't limit search / select in specific scope, or with a single command to select all matches in the scope
11. Can't run test inside editor
12. Missing some wrappers or tools to work with `Cargo` without going into the complete command way / can't run `Cargo build` or other command inside the editor, have to open terminal first
13. Can't click or go to the file position like `--> src/bin/some_bin.rs:123:24` when compile failed
14. Completion candidates sorting not ideal


I have all these things in Emacs. But still,  Zed is blazingly fast and I'm quite happy with it.

**Comment by Manubi** [comment]
![CleanShot 2024-01-30 at 20 36 41](https://github.com/zed-industries/zed/assets/20935138/103e8d28-5a6c-46a7-9e25-ce72675ac1b1)
I know it's probably not possible, but i would be cool to have plugin or whatever which let's me define a test case for a function, injects the test case input into the function which i am writing and during I am writing the function i get already the current state of the updated input. 
So I basically see what every part does. like the .split the .take(1) and so on. Would be a massive help for beginners which would lead to a USP for zed.

**Reply by endigma** [reply]
You can probably just do something like:

```rs
trait InspectValue {
    fn inspect_value(self, label: &str) -> Self;
}

impl<T: std::fmt::Debug> InspectValue for T {
    fn inspect_value(self, label: &str) -> Self {
        println!("inspected {}, got {:?}", label, self);
        self
    }
}
```

Or maybe a macro, obviously I wouldn't recommend leaving this in for your finished product.

https://www.rustexplorer.com/b/ue6oy3

**Comment by franciskafyi** [comment]
I'm closing this Discussion as stale.
